packages/rpc-graphql/src/loaders/account.ts

```diff
@@ -29,14 +28,36 @@ function normalizeArgs({
 
 /* Load an account from the RPC, transform it, then return it */
 async function loadAccount(rpc: Rpc, { address, ...config }: ReturnType<typeof normalizeArgs>) {
+    if (
+        address === 'Feature111111111111111111111111111111111111' ||
+        address === 'NativeLoader1111111111111111111111111111111' ||
+        address === 'Sysvar1nstructions1111111111111111111111111'
+    ) {
+        // This is a synthetic account; `getAccountInfo()` would return `null`.
+        return {
+            address,
+            data: ['', 'base64'],
+            executable: false,
+            lamports: 0n,
+            owner: '11111111111111111111111111111111',
+            rentEpoch: 18_446_744_073_709_552_000n,
+            space: 0n,
+        };
+    }
     const account = await rpc
         .getAccountInfo(address, config as Parameters<SolanaRpcMethods['getAccountInfo']>[1])
         .send()
         .then(res => res.value)
         .catch(e => {
             throw e;
         });
-    return account === null ? { address } : transformLoadedAccount({ account, address, encoding: config.encoding });
+    if (!account || account.lamports <= 0n) {
+        return null;
+    }
+    return {
+        ...account,
+        address,
+    };
```

Comment:
I _want_ it to return `null`.

Account has no rent?
It doesn't exist.
Return `null`.

Reply:
I haven't really used rpc-graphql so LMK if this doesn't make sense, but I'm wondering if we should align with the `MaybeAccount` structure Loris added for `@solana/accounts`?

That'd be something like this when the account exists:
```
  ...account,
  address,
  exists: true
```

And this when it doesn't:
```
  address,
  exists: false
```

Reply:
I think I know what the use case for this is, but check me on this. Someone could make this query:

```graphql
{
  transaction(signature: $signature) {
    instructions {
      accounts {
        address
      }
    }
  }
}
```

…and if over time one of those accounts had been deleted (rent, withdrawn), you'd end up with this:

```json
{
  "transaction": {
    "instructions: [{
      "accounts: [
        {address: "foo"},
        null,
        {address: "baz"},
      ]
    }]
  }
}
```

Is that the use case wer're trying to solve for here? You fetch an account that's since gone away, but the purpose of the query is such that you just wanted to know (at least) the address of the account at that point in time?

Reply:
> I think I know what the use case for this is, but check me on this. Someone could make this query:

I would say if you request anything that's already in the returned layer 1 object then there's no need to transform or load anything. So in this example, `instructions.accounts` already has the addresses, even if the accounts themselves don't exist anymore. If I'm only interested in the addresses, there's no need to make any calls, and I should just get the addresses:

```json
{
  "transaction": {
    "instructions": [{
      "accounts": [
        {"address": "foo"},
        {"address": "foo2"},
        {"address": "baz"},
      ]
    }]
  }
}
```

Once you add any other field that *can only be obtained by an RPC call*, that's when you should see a `null` or not. In @mcintyre94's suggestion, you'd maybe do:

```gql
{
  transaction(signature: $signature) {
    instructions {
      accounts {
        address
        exists
      }
    }
  }
}
```

And then instead of (which wouldn't make sense):

```json
{
  "transaction": {
    "instructions": [{
      "accounts": [
        {"address": "foo", "exists": true },
        null,
        {"address": "baz", "exists": true },
      ]
    }]
  }
}
```

You'd get:

```json
{
  "transaction": {
    "instructions": [{
      "accounts": [
        {"address": "foo", "exists": true },
        {"address": "foo2", "exists": false },
        {"address": "baz", "exists": true },
      ]
    }]
  }
}
```

Side note: admittedly I believe the whole damn thing goes `null` if `graphql` can't build out the proper response, so I think it would also be beneficial to add support for sticking a `null` field-wise in the response, so that doesn't happen.

Reply:
To be more specific, in the current implementation, for account `"foo2"` that doesn't exist, this query:

```gql
{
  transaction(signature: $signature) {
    instructions {
      accounts {
        address
        lamports
      }
    }
  }
}
```

...would return:

```json
{
  "transaction": null
}
```

...which is bad, bad. Instead you should get:

```json
{
  "transaction": {
    "instructions": [{
      "accounts": [
        {"address": "foo", "lamports": 500n },
        {"address": "foo2", "lamports": null },
        {"address": "baz", "lamports": 500n },
      ]
    }]
  }
}
```

But a case could be made for some other way to let you know the account itself is `null` (doesn't exist), not just the `lamports` field.

Reply:
With `MaybeAccount` it's mostly about resolving the fact that our RPC calls don't return the address.

So I think I agree with Joe, if you pass `address` as a requested return field, then we should return it even if the account doesn't exist. Otherwise you have to keep your requested addresses somewhere and match on the index which is a bit annoying.

I guess that just leaves the question of whether we should have an `exists` field that can be requested. You probably need some way of determining that because some fields are genuinely nullable, and you might need to distinguish between null field and account not existing if you get a null there. But I guess every field has `lamports` so comparing that to `null` is a good solution.

I think I'd still prefer to add `exists` because I see this as being about as high-level as `@solana/accounts`, that's what we do there, and it's better type safety, but it's quite minor. If you can return `{"address": "foo2", "lamports": null },` you can build whatever you need based on that.

Reply:
# First, some background.

GraphQL, because it's basically just a type contract between a server and client, naturally lends itself to returning tagged unions. This is amazing for clients, because they can do things like this:

```ts
const account = await fetchWithGraphQL(graphql`
  query AccountQuery($address: Address!) {
    account(address: $address) {
      __typename
      lamports
      ... on TokenAccount {
        mint
      }
      ... on VoteAccount {
        authority
      }
    }
  }
`, {address: "abc"});

account satisfies (null | Account);

if (account == null) {
  throw new Error('I know "abc" does not exist');
}

account satisfies Account;  // We know it to be non-null now (existent)

 // TypeScript can *actually* refine the type based on what's in `__typename`
switch (account.__typename) {
  case 'TokenAccount': {
    account satisfies TokenAccount;
    account.mint satisfies Address;
    console.log('"abc" is a token account whose mint is', account.mint);
  }
  case 'VoteAccount': {
    account satisfies VoteAccount;
    account.authority satisfies Address;
    console.log('"abc" is a vote account whose authority is', account.authority);
  }
  default: {
    account.lamports satisfies bigint;
    console.log('whatever kind of account "abc" is, it has this balance', account.lamports);
  }
}
```

Semantics game: on point. If we instead contextualized ‘an account existing’ as ‘its `exists` property being `true`’:

```ts
const account = /* same as above */

if (!account.exists) {
  throw new Error('I know "abc" does not exist');
}

// TypeScript can't refine these values for shit.
account.lamports satisfies bigint; // FAIL: it's still (bigint | undefined)
account.space satisfies bigint; // FAIL: it's still (bigint | undefined)
```

# Now, if I can back way up for a second.

There are two use cases for fetching accounts that I can see.

1. You know the address of the account.
2. You don't.

## Example of case 1

Fetching the balance of an account.

```graphql
query {
  account(address: $address) {
    lamports
  }
}
```

The frontend code for this _already knows_ what the address is. If the account doesn't exist, the query should null out or throw, and the UI code looks like this:

```ts
function AccountBalance({address}) {
  const {account} = await fetchWithGraphQL(gql`
    query AccountBalanceQuery($address: Address!) {
      account(address: $address) @required(action: THROW) {
        lamports
      }
    }
  `, {address});
  // If the code reaches this point, the `account` and the `lamports` field is *guaranteed* to exist.
  return <span>{account.lamports} SOL</span>;
}

function App() {
  const wallet = useWallet();
  return (
    <ErrorBoundary fallback={<span>Wallet {wallet.address} does not exist</span>}>
      Wallet {wallet.address} balance:
      {' '}
      <React.Suspense fallback={<LoadingSpinner />}>
        <AccountBalance address={wallet.address} />
      </React.Suspense>
    </ErrorBoundary>
  );
}
```

This code would be _significantly_ more wrought and error prone if ‘knowing that the account does not exist’ relied on runtime `false` checks of `exists`.

# Example of case 2

OK here's where things get hard. Sometimes both of these are true:

* you are interested in an address
* there is no account at that address

This can happen in the case of fetching historical transactions:

```graphql
query {
  transaction(signature: $signature) {
    instructions {
      accounts {
        address
      }
    }
  }
}
```

Shit, what if the account used in that transaction no longer exists?

```json
{
  "transaction": [
    "instructions": [
      "accounts": [
        {"address": "abc"},
        null, // Shit
        {"address": "ghi"},
      ]
    ]
  ]
}
```

This is the only example I can think of where you don't know the address of an account, need to know it, and it no longer exists.

# So what do we do about case 2?

My example is bad, because that's not actually how `accounts` work on `instructions`. They're actually this:

```json
{
  "address": "abc",
  "role": "WRITABLE"
}
```

Let's imagine though that you were building an explorer and you _did_ want the ability to render a whole-ass interface that showed the balance of an account _inside the instructions viewer_. We could do this:

```graphql
type InstructionAccount {
  address: Address!
  account: Account
  role: AccountRole
}
```

Then your query could be:

```graphql
{
  transaction(signature: $signature) {
    instructions {
      accounts {
        address
        account {
          lamports
        }
        role
      }
    }
  }
}
```

Now the query is usable for whatever purpose.

```json
{
  "transaction": [
    "instructions": [
      "accounts": [
        {"address": "abc", "account": {"lamports": 1n}, "role": "WRITABLE_SIGNER"},
        {"address": "def", "account": null, "role": "WRITABLE"},
        {"address": "ghi", "account": {"lamports": 324123n}, "role": "READONLY"},
      ]
    ]
  ]
}
```

Your instructions-rendering UI could just display like ‘hey here lies account `"def"` that was once writable but now doesn't exist, RIP.’

> [!NOTE]
> Maybe, as the explorer developer, you want to display the _historical_ balance at the time the instruction was landed. Well, too bad, out of scope for the Solana RPC. Build your own lookup.

# Side note

Just to get ahead of it, people will _for sure_ want this kind of an interface for `InstructionAccount`:

```graphql
type InstructionAccount {
  address: Address!
  account: Account
  tokenBalances {
    mint {
      address
      account {
        authority
      }
    }
    pre
    post
  }
  role: AccountRole
}
```

This is one of many of our chances to destructure RPC response data from where it is, to where it belongs, in the mind of a UI developer.

Reply:
Nice, thanks for the context and great explanation. I agree with everything you've laid out, however that leads me to ask about this particular use case.

I know the address I'm querying because I have it on my frontend, but I *don't* know the address of this account's owner. My query seeks to fetch information about *both* accounts.

We've discussied doing this:

```gql
query {
  account(address: "abc") {
    lamports
    owner {
      address
      lamports
    }
  }
}
```

...and you'd get this if the owner account doesn't exist:

```json
{
  "account": {
    "lamports": 500,
    "owner": {
      "address": "1111...",
      "lamports": null
    }
  }
}
```

However, based on what you've written above, we can instead do this:

```gql
query {
  account(address: "abc") {
    lamports
    owner {
      address
      account {
        lamports
      }
    }
  }
}
```

...and you'd get this if the owner account doesn't exist:

```json
{
  "account": {
    "lamports": 500,
    "owner": {
      "address": "1111...",
      "account": null
    }
  }
}
```

That feels a little nicer, but it does add some overhead to our schema for querying typed accounts in those areas:

```gql
query {
  account(address: "abc") {
    lamports
    owner {
      address
      account {
        ... on MintAccount {
          decimals
          isInitialized
        }
      }
    }
  }
}
```

...and you'd get this if the owner account doesn't exist:

```json
{
  "account": {
    "lamports": 500,
    "owner": {
      "address": "1111...",
      "account": {
        "decimals": 0,
        "isInitialized": true
      }
    }
  }
}
```

The current implementation allows for this - one less layer on the spread type:

```gql
query {
  account(address: "abc") {
    lamports
    owner {
      address
      ... on MintAccount {
        decimals
        isInitialized
      }
    }
  }
}
```

Reply:
Damn. One thing I meant to claim in my message above is that I think we should _remove_ `address` from the `Account` interface.

> I know the address I'm querying because I have it on my frontend, but I don't know the address of this account's owner.

By definition, if an account doesn't exist, it's ‘owned’ by the System Program, right? You can query `account.owner.owner.owner.owner` forever, until you hit a `null`, at which point you know that, by definition, the owner of the non-existent (`null`) account is the System Program.

That's enough, isn't it?

Reply:
> ...and you'd get this if the owner account doesn't exist:

Oh, wait. Is the claim here that you can have an account that's owned by some program account, but that program account has since been deallocated (`lamports` withdrawn), such that if we return `null` (because it no longer exists) you can't actually learn what the owner's address is.

Hmm…

Reply:
Yeah, this smells like wherever that's necessary, we need the wrapper type.

```graphql
type InstructionAccount {
  address: Address!
  account: Account
}

type OwningProgram {
  address: Address!
  account: ProgramAccount
}

type Mint {
  address: Address!
  account: MintAccount
}
```

I don't hate it.

Reply:
In particular, I don't hate it, because in this example:

```graphql
query {
  account(address: "abc") {
    lamports
    owner {
      address
      ... on MintAccount {
        decimals
        isInitialized
      }
    }
  }
}
```

`owner` can never actually be a `MintAccount`. Owner is always a `ProgramAccount`.